### PR TITLE
planner: reduce topn count to exclude non-skewed values | tidb-test=pr/2333

### LINKED ISSUE #53111: TopN statistics are collected unnecessarily for non-skewed values on small tables
## Enhancement

Current ANALYZE will collect by default the top 500 values for an interesting column (indexed column or predicate column). For tables that are sampled, then there is pruning logic to remove statistics that are not skewed. For tables that are NOT sampled (smaller tables) - the pruning logic is not invoked.

Below shows an example where values are collected with a count of 1 - which are not skewed.

```
tidb> show stats_topn;
+---------+------------+----------------+-------------+----------+------------+-------+
| Db_name | Table_name | Partition_name | Column_name | Is_index | Value      | Count |
+---------+------------+----------------+-------------+----------+------------+-------+
| test    | t2         |                | a           |        0 | 73         |    16 |
| test    | t2         |                | a           |        0 | 74         |    16 |
| test    | t2         |                | a           |        0 | 75         |    16 |
| test    | t2         |                | a           |        0 | 76         |    16 |
......
| test    | t2         |                | a           |        0 | 101        |     1 |
| test    | t2         |                | a           |        0 | 102        |     1 |
| test    | t2         |                | a           |        0 | 103        |     1 |
| test    | t2         |                | a           |        0 | 104        |     1 |
......
```

### PR DESCRIPTION
**Pull request by terry1purcell**
### What problem does this PR solve?

Issue Number: close #53111 

Problem Summary:

### What changed and how does it work?

PR includes 3 changes to reduce unnecessary TopN values:
1. Reduce the default TopN to 100 (from 500).
2. Support pruning of TopN values that are "not skewed" for smaller tables. 
3. And initially avoid adding values with a count=1 (which will be pruned anyway).

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,124 +0,0 @@
-<div align="center">
```
Comment (elsa0520) [inline README.md:1]: Maybe PR shouldn't remove the README.md  ?

File: pkg/planner/core/planbuilder.go [hunk 2]
```diff
@@ -2684,7 +2684,7 @@ var analyzeOptionDefault = map[ast.AnalyzeOptionType]uint64{
 
 var analyzeOptionDefaultV2 = map[ast.AnalyzeOptionType]uint64{
 	ast.AnalyzeOptNumBuckets:    256,
-	ast.AnalyzeOptNumTopN:       500,
```
Comment (elsa0520) [inline pkg/planner/core/planbuilder.go:2687]: Could you please explain why we changed the TopNNum from 500 to 100 in comments ?

File: pkg/statistics/handle/globalstats/global_stats_internal_test.go [hunk 3]
```diff
@@ -351,6 +351,7 @@ func testIssues24349(testKit *testkit.TestKit) {
 	testKit.MustExec("create table t (a int, b int) partition by hash(a) partitions 3")
 	testKit.MustExec("insert into t values (0, 3), (0, 3), (0, 3), (0, 2), (1, 1), (1, 2), (1, 2), (1, 2), (1, 3), (1, 4), (2, 1), (2, 1)")
 	testKit.MustExec("analyze table t with 1 topn, 3 buckets")
```
Comment (elsa0520) [inline pkg/statistics/handle/globalstats/global_stats_internal_test.go:353]: Should we add some new test about the pruning logic in smaller table ? And check the topn number which has been pruned ?

File: pkg/planner/core/planbuilder.go (hunk 2 above)
Comment (terry1purcell) [reply pkg/planner/core/planbuilder.go:2687]: Added comment as requested

File: pkg/statistics/handle/globalstats/global_stats_internal_test.go (hunk 3 above)
Comment (terry1purcell) [reply pkg/statistics/handle/globalstats/global_stats_internal_test.go:353]: Added a testcase to TestPruneTopN function in statistics_test.go

File: README.md (hunk 1 above)
Comment (terry1purcell) [reply README.md:1]: I attempted to remove it from the PR, but wasn't intending to remove it from the pingCAP's repo! Let me investigate.

Comment (terry1purcell) [reply README.md]: I corrected this error

**Review by elsa0520 — APPROVED** [review]
LGTM

File: pkg/statistics/builder.go [hunk 4]
```diff
@@ -454,8 +466,7 @@ func BuildHistAndTopN(
 //
 //	We assume that the ones not in the top-n list's selectivity is 1/remained_ndv which is the internal implementation of EqualRowCount
 func pruneTopNItem(topns []TopNMeta, ndv, nullCount, sampleRows, totalRows int64) []TopNMeta {
-	// If the sampleRows holds all rows, or NDV of samples equals to actual NDV, we just return the TopN directly.
-	if sampleRows == totalRows || totalRows <= 1 || int64(len(topns)) >= ndv || len(topns) == 0 {
```
Comment (winoros) [inline pkg/statistics/builder.go:458]: @time-and-fate I think you need to take a look at this line's change

Comment (time-and-fate) [reply pkg/statistics/builder.go:458]: This looks good to me.
I checked the previous bad case and I believe the `int64(len(topns)) >= ndv` check should prevent that.
As for the change from `len(topns) == 0` to `len(topns) <= 1`, it looks like this won't bring a big behavior change.

Comment (time-and-fate) [reply pkg/statistics/builder.go:458]: But we may still need to be ready for possible regressions reported from the daily benchmarks or release tests.

File: pkg/statistics/builder.go [hunk 5]
```diff
@@ -360,7 +370,7 @@ func BuildHistAndTopN(
 
 	// Handle the counting for the last value. Basically equal to the case 2 above.
 	// now topn is empty: append the "current" count directly
-	if numTopN != 0 {
+	if numTopN != 0 && (!allowPruning || (allowPruning && (sampleFactor <= 1 || curCnt > 1))) {
```
Comment (hi-rustin) [inline pkg/statistics/builder.go:373]: Maybe we should add some comments to explain here we still try to ignore topn with count 1.

File: pkg/statistics/statistics_test.go [hunk 6]
```diff
@@ -674,13 +674,25 @@ func TestPruneTopN(t *testing.T) {
...
+	topnIn = []TopNMeta{
+		{[]byte{1}, 3_000},
+		{[]byte{2}, 3_000},
+	}
+	totalNDV = 4002
+	nullCnt = 0
+	sampleRows = 10_000
+	totalRows = 10_000
+	topnOut = pruneTopNItem(topnIn, totalNDV, nullCnt, sampleRows, totalRows)
+	require.Equal(t, topnIn, topnOut)
```
Comment (hi-rustin) [inline pkg/statistics/statistics_test.go:697]: Do we need to add a case to check if we can prune `1` correctly?

**Review by hi-rustin — COMMENTED** [review]
Thanks!

Some nits.

**Review by hi-rustin — APPROVED** [review]
Approve in advance.

Feel free to `/unhold` the PR after adding the test and comment.

/hold

**Review by hi-rustin — APPROVED** [review]
Looks good to me. Thanks! 👍

**Pull request merged**